src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ConstructorParametersShouldMatchPropertyAndFieldNames.cs

```diff
@@ -0,0 +1,261 @@
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+
+            context.RegisterCompilationStartAction(
+                (compilationStartContext) =>
+                {
+                    INamedTypeSymbol? jsonConstructorAttributeNamedSymbol = compilationStartContext.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.SystemTextJsonSerializationJsonConstructorAttribute);
+                    if (jsonConstructorAttributeNamedSymbol == null)
+                    {
+                        return;
+                    }
+
+                    var paramAnalyzer = new ParameterAnalyzer(jsonConstructorAttributeNamedSymbol);
+
+                    compilationStartContext.RegisterOperationBlockStartAction(
+                        (startOperationBlockContext) =>
+                        {
+                            if (startOperationBlockContext.OwningSymbol is IMethodSymbol method
+                                && !paramAnalyzer.ShouldAnalyzeMethod(method))
+                            {
+                                return;
+                            }
+
+                            startOperationBlockContext.RegisterOperationAction(
+                                context => ParameterAnalyzer.AnalyzeOperationAndReport(context),
+                                OperationKind.ParameterReference);
+                        });
+                });
+        }
+
+        internal enum ParameterDiagnosticReason
+        {
+            NameMismatch,
+            PropertyInappropriateVisibility
+        }
+
+        private sealed class ParameterAnalyzer
+        {
+            private readonly INamedTypeSymbol _jsonConstructorAttributeInfoType;
+
+            public ParameterAnalyzer(INamedTypeSymbol jsonConstructorAttributeInfoType)
+            {
+                _jsonConstructorAttributeInfoType = jsonConstructorAttributeInfoType;
+            }
+
+            public static void AnalyzeOperationAndReport(OperationAnalysisContext context)
+            {
+                var operation = (IParameterReferenceOperation)context.Operation;
+
+                IMemberReferenceOperation? memberReferenceOperation = TryGetMemberReferenceOperation(operation);
+                ISymbol? referencedSymbol = memberReferenceOperation?.GetReferencedMemberOrLocalOrParameter();
+
+                if (referencedSymbol == null || referencedSymbol.IsStatic)
+                {
+                    return;
+                }
+
+                IParameterSymbol param = operation.Parameter;
+                var field = referencedSymbol as IFieldSymbol;
+                var prop = referencedSymbol as IPropertySymbol;
+
+                if (field == null && prop == null)
+                {
+                    return;
+                }
+
+                if (IsSupportedField(field) && !IsParamMatchFieldName(param, field))
```

Comment:
I'd do something like:

```csharp
if (referencedSymbol is IFieldSymbol field)
{
    if(!IsParamMatchFieldName(.....)
    {
        // DoWork
    }
}
else if (referencedSymbol is IPropertySymbol prop)
{
    // Do checks and work
}
```

Reply:
Yep, it makes sense. If additional checks appear later for supported fields than it's easy to convert it back to methods. Done.